### PR TITLE
fix the `alpha` of focal loss

### LINKED ISSUE #875: Bug about the weight factor alpha of focal loss
## 🐛 Bug

https://github.com/kornia/kornia/blob/cf8e854546110e3dac1caa35449429fc7af328e5/kornia/losses/focal.py#L86

Hi, I find someting wrong about the multi-class focal loss computation.

It seems like we are multiplying the loss of different classes by a same alpha factor. I think this is equal to just reduce lr, and cannot address the class imbalance issue. In the original paper, alpha is defined as another cased function just like class probability (Eq.2). 

Please take a look about it, thank you !

### PR DESCRIPTION
**Pull request by qingpeng9802**
Fixes #875 (issue)

#### Changes
1. fix the `alpha` of focal loss in softmax version.
2. add `None` for `alpha`. Although `alpha=0.5` has the equivalent training effect as `alpha=None`, some users might need a value from non-alpha balanced version.
3. benchmark shows that the `torch.eisum` is actually ~25% slower than `torch.sum` + `*` so the `torch.eisum` is replaced (see https://github.com/kornia/kornia/pull/1362). Might need discussion, see https://github.com/kornia/kornia/pull/2372
4. the algorithm formula is changed to use `logsigmoid` and `log_softmax` as much as possible. This might be able to improve the numerical stability potentially as the idea in https://github.com/kornia/kornia/pull/1362.
5. add `weight` for `focal_loss` and `binary_focal_loss_with_logits`
6. remove `eps`
7. result tensor of `focal_loss`  is no longer `.sum(1)`ed before reduction
8. refactor the unit tests to improve coverage
9. update inline doc for consistency

breaking changes:
1, 6, 7

#### Reference
a small benchmark snippet to who may need
```python
i = torch.rand(10, 20, 1000, 1000).cuda()
t = torch.randint(0,20,(10, 1000, 1000)).cuda()
tt = F.one_hot(t, num_classes=-1).permute(0,3,1,2).contiguous()
def do_self():
    for _ in range(100):
        focal_loss(i, t, alpha=0.5)
    torch.cuda.synchronize()
import time
torch.cuda.synchronize()
start = time.perf_counter()
do_self()
print(time.perf_counter()-start)
```

implementation reference is from https://github.com/pytorch/pytorch/blob/6e0c741105b4022fc7c3f8c2919aad9f971beb62/modules/detectron/softmax_focal_loss_op.cu#L59-L85

#### Type of change

- [x] 📚  Documentation Update
- [x] 🧪 Tests Cases
- [x] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [x] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

File: test/integration/test_focal.py [hunk 1]
```diff
@@ -19,7 +19,7 @@ class TestIntegrationFocalLoss:
     num_classes = 2
 
     # focal loss
-    alpha = 2.0
+    alpha = 0.5
```
Comment (qingpeng9802) [inline test/integration/test_focal.py:22]: old `alpha` is actually not working. To get the same result as before, `alpha` should be balanced, that is, `0.5`.

File: kornia/losses/focal.py [hunk 2]
```diff
@@ -205,16 +227,20 @@ def binary_focal_loss_with_logits(
...
+    if alpha is None:
+        loss_tmp = (
+            -pos_weight * log_probs_neg.exp().pow(gamma) * target * log_probs_pos
+            - log_probs_pos.exp().pow(gamma) * (1.0 - target) * log_probs_neg
+        )
+    else:
+        loss_tmp = (
+            -alpha * pos_weight * log_probs_neg.exp().pow(gamma) * target * log_probs_pos
+            - (1.0 - alpha) * log_probs_pos.exp().pow(gamma) * (1.0 - target) * log_probs_neg
+        )
```
Comment (qingpeng9802) [inline kornia/losses/focal.py:243]: if the alpha term is extracted as 
```python
if alpha is not None:
    alpha_fac=alpha*target+(1.-alpha)*(1.-target)
```
benchmark shows this is ~20% slower.

File: kornia/losses/focal.py [hunk 3]
```diff
@@ -68,19 +75,30 @@ def focal_loss(
...
+                f"`alpha` shape must be (num_of_classes,), that is, {(num_of_classes,)}. Got {alpha.shape}"
+            )
+        else:
+            raise TypeError(f'Expected the `alpha` be a float | Tensor | None. Got {type(alpha)}.')
+
+        boradcast_dims = [-1] + [1] * len(input.shape[2:])
+        alpha_fac = alpha_fac.view(boradcast_dims).to(loss_tmp)
+        loss_tmp = loss_tmp * alpha_fac
+
+    loss_tmp = loss_tmp.sum(1)
```
Comment (qingpeng9802) [inline kornia/losses/focal.py:101]: See Changes 3.

File: kornia/losses/focal.py [hunk 4]
```diff
@@ -68,19 +75,30 @@ def focal_loss(
...
-    loss_tmp = torch.einsum('bc...,bc...->b...', (target_one_hot, focal))
+    # compute softmax over the classes axis
+    log_input_soft: Tensor = input.log_softmax(1)
 
+    # compute the actual focal loss
+    loss_tmp = -torch.pow(1.0 - log_input_soft.exp(), gamma) * log_input_soft * target_one_hot
+    if alpha is not None:
+        num_of_classes = input.shape[1]
+        if isinstance(alpha, float):
+            alpha_fac = torch.tensor([1 - alpha] + [alpha] * (num_of_classes - 1))
```
Comment (edgarriba) [inline kornia/losses/focal.py:89]: pass device and dtype

Comment (edgarriba) [inline kornia/losses/focal.py:89]:
```suggestion
            alpha_fac = tensor([1 - alpha] + [alpha] * (num_of_classes - 1))
```
`from kornia.core import tensor`

File: kornia/losses/focal.py [hunk 5]
```diff
@@ -68,19 +75,30 @@ def focal_loss(
         f"input and target must be in the same device. Got: {input.device} and {target.device}",
     )
```
Comment (johnnv1) [inline kornia/losses/focal.py:77]:
```suggestion
        KORNIA_CHECK_TYPE(alpha, (float, Tensor, None))
```

File: kornia/losses/focal.py [hunk 6]
```diff
@@ -205,16 +227,20 @@ def binary_focal_loss_with_logits(
...
-
     log_probs_pos = nn.functional.logsigmoid(input)
     log_probs_neg = nn.functional.logsigmoid(-input)
 
-    loss_tmp = (
-        -alpha * pos_weight * probs_neg.pow(gamma) * target * log_probs_pos
-        - (1 - alpha) * probs_pos.pow(gamma) * (1.0 - target) * log_probs_neg
-    )
+    # the alpha term is not extracted to save operations
+    if alpha is None:
```
Comment (johnnv1) [inline kornia/losses/focal.py:234]: Can you add tests for both cases?

File: kornia/losses/focal.py [hunk 7]
```diff
@@ -68,19 +75,30 @@ def focal_loss(
...
+    if alpha is not None:
+        num_of_classes = input.shape[1]
+        if isinstance(alpha, float):
+            alpha_fac = torch.tensor([1 - alpha] + [alpha] * (num_of_classes - 1))
+        elif isinstance(alpha, Tensor) and alpha.shape != (num_of_classes,):
+            raise ValueError(
+                f"`alpha` shape must be (num_of_classes,), that is, {(num_of_classes,)}. Got {alpha.shape}"
+            )
+        else:
+            raise TypeError(f'Expected the `alpha` be a float | Tensor | None. Got {type(alpha)}.')
```
Comment (johnnv1) [inline kornia/losses/focal.py:95]: KORNIA_CHECK_SHAPE -- from kornia.core.check import KORNIA_CHECK_SHAPE

```suggestion
        KORNIA_CHECK_SHAPE(alpha, [str(num_of_classes),], f"`alpha` shape must be (num_of_classes,)!")
```

**Comment by qingpeng9802** [comment]
I noticed that when the `alpha` in the softamx version is used as a class weight, users may also have the same need in the sigmoid version. From the perspetive of the symmetry of interface parameters, do we consider splitting the parameter `alpha` into `alpha: float | None` and `weight: Tensor | None`? Then, we can add `weight: Tensor | None` for both `focal_loss` and `binary_focal_loss_with_logits`. 
@edgarriba any comment about this?

File: kornia/losses/focal.py (hunk 6 above)
Comment (qingpeng9802) [reply kornia/losses/focal.py:234]: Sure, let me set the new design first, and more tests will be added later.

**Review by edgarriba — APPROVED** [review]
LGTM. @johnnv1 final review and merge

**Comment by johnnv1** [comment]
> I noticed that when the `alpha` in the softamx version is used as a class weight, users may also have the same need in the sigmoid version. From the perspetive of the symmetry of interface parameters, do we consider splitting the parameter `alpha` into `alpha: float | None` and `weight: Tensor | None`? Then, we can add `weight: Tensor | None` for both `focal_loss` and `binary_focal_loss_with_logits`. @edgarriba any comment about this?

i think is reasonable for this to be alpha and weight -- as I believe this will change the way it currently works, I advocate adding a warning that we are switching to this way, and somehow keep the old functionality for at least 1 release with warning

File: test/losses/test_losses.py [hunk 8]
```diff
@@ -77,7 +77,7 @@ def test_dynamo(self, device, dtype, torch_optimizer):
     def test_gradcheck(self, device, dtype):
         alpha, gamma = 0.5, 2.0  # for focal loss with logits
         logits = torch.rand(2, 3, 2).to(device, dtype)
-        labels = torch.rand(2, 1, 3, 2)
+        labels = torch.rand(2, 3, 2)
```
Comment (qingpeng9802) [inline test/losses/test_losses.py:80]: historical issue, not matching shape

File: kornia/losses/focal.py [hunk 9]
```diff
@@ -194,27 +232,51 @@ def binary_focal_loss_with_logits(
...
 
-    if pos_weight is None:
-        pos_weight = torch.ones([input.shape[1], *([1] * len(input.shape[2:]))], device=input.device, dtype=input.dtype)
-
-    KORNIA_CHECK_IS_TENSOR(pos_weight)
-    KORNIA_CHECK(input.shape[1] == pos_weight.shape[0], "Expected pos_weight equals number of classes.")
+    log_probs_pos: Tensor = nn.functional.logsigmoid(input)
+    log_probs_neg: Tensor = nn.functional.logsigmoid(-input)
+
+    pos_term: Tensor = -log_probs_neg.exp().pow(gamma) * target * log_probs_pos
```
Comment (qingpeng9802) [inline kornia/losses/focal.py:244]: Splitting original into two terms should have no performance issue here. The benchmark shows no performance decrease.

File: kornia/losses/focal.py [hunk 10]
```diff
@@ -14,7 +14,13 @@
...
 
 def focal_loss(
-    input: Tensor, target: Tensor, alpha: float, gamma: float = 2.0, reduction: str = 'none', eps: float | None = None
+    input: Tensor,
+    target: Tensor,
+    alpha: float | None,
+    gamma: float = 2.0,
+    reduction: str = 'none',
+    eps: float | None = None,
+    weight: Tensor | None = None,
```
Comment (qingpeng9802) [inline kornia/losses/focal.py:23]: This way to add `weight` should not have any backward compatibility issues.

**Comment by qingpeng9802** [comment]
Two more things about design:
* Do we have any plans to remove the `eps` parameter, since the `eps` warning has been shown for 21 months?
*  Currently, `focal_loss` returns a tensor that is not compatible with the pytorch `none` style, that is, the tensor is summed over Channel dim before return. This may be a bit confusing for users, see a similar issue in another repo https://github.com/Project-MONAI/MONAI/issues/3185 . My proposal is adding a mode `torch-none`, and the mode returns a pytorch `none` style tensor. Any comment about this?

**Comment by qingpeng9802** [comment]
> Two more things about design:
> 
> * Do we have any plans to remove the `eps` parameter, since the `eps` warning has been shown for 21 months?
> * Currently, `focal_loss` returns a tensor that is not compatible with the pytorch `none` style, that is, the tensor is summed over Channel dim before return. This may be a bit confusing for users, see a similar issue in another repo [Focal loss with reduction=none returns an incorrect shape Project-MONAI/MONAI#3185](https://github.com/Project-MONAI/MONAI/issues/3185) . My proposal is adding a mode `torch-none`, and the mode returns a pytorch `none` style tensor. Any comment about this?

@edgarriba @johnnv1 Any comment about this?

**Comment by johnnv1** [comment]
> I'm in favor to follow pytorch convention. Probably other losses have to be updated too. Removing the eps LGTM

+1

**Comment by edgarriba** [comment]
@qingpeng9802 @johnnv1 can we merge this ?

**Comment by qingpeng9802** [comment]
> @qingpeng9802 @johnnv1 can we merge this ?

Sure. As mentioned above, the return of `focal_loss` now follows pytorch convention, and this breaking change may need to be mentioned in the next release.
